actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -46,8 +49,168 @@ class TagBuilder # :nodoc:
         include CaptureHelper
         include OutputSafetyHelper
 
-        HTML_VOID_ELEMENTS = %i(area base br col embed hr img input keygen link meta param source track wbr).to_set
-        SVG_SELF_CLOSING_ELEMENTS = %i(animate animateMotion animateTransform circle ellipse line path polygon polyline rect set stop use view).to_set
+        def self.define_element(name, code_generator:, method_name: name.to_s.underscore)
+          code_generator.define_cached_method(method_name, namespace: :tag_builder) do |batch|
+            batch.push(<<~RUBY) unless instance_methods.include?(method_name.to_sym)
+              def #{method_name}(*arguments, **options, &block)
+                tag_string(#{name.inspect}, *arguments, **options, &block)
+              end
+            RUBY
+          end
+        end
+
+        def self.define_void_element(name, code_generator:, method_name: name.to_s.underscore, self_closing: false)
+          code_generator.define_cached_method(method_name, namespace: :tag_builder) do |batch|
+            batch.push(<<~RUBY)
+              def #{method_name}(content = nil, escape: true, **options, &block)
+                if content.nil? && block.nil?
```

Comment:
```suggestion
                if content.nil? && block
```

Ok, so this is a very subtle optimization regarding `&block`.

There are a subset of operations you can do with `&block` that will use the `blockparamproxy`, this is very cheap.

If you do any unsupported operation with the block, then instead of the proxy, Ruby will have to allocate a `Proc` object which is costly.

In this case `block.nil?` cause the allocation.

```
>> puts RubyVM::InstructionSequence.compile("def foo(&block)\nputs :hey if true && block.nil? \nend").disasm
== disasm: #<ISeq:<compiled>@<compiled>:1 (1,0)-(3,3)> (catch: false)
0000 definemethod                           :foo, foo                 (   1)[Li]
0003 putobject                              :foo
0005 leave

== disasm: #<ISeq:foo@<compiled>:1 (1,0)-(3,3)> (catch: false)
local table (size: 1, argc: 0 [opts: 0, rest: -1, post: 0, block: 0, kw: -1@-1, kwrest: -1])
[ 1] block@0<Block>
0000 getblockparam                          block@0, 0                (   2)[LiCa]
0003 opt_nil_p                              <calldata!mid:nil?, argc:0, ARGS_SIMPLE>[CcCr]
0005 branchunless                           13
0007 putself
0008 putobject                              :hey
0010 opt_send_without_block                 <calldata!mid:puts, argc:1, FCALL|ARGS_SIMPLE>
0012 leave                                                            (   3)[Re]
0013 putnil                                                           (   2)
0014 leave                                                            (   3)[Re]
=> nil
```
vs

```
>> puts RubyVM::InstructionSequence.compile("def foo(&block)\nputs :hey if true && block \nend").disasm
== disasm: #<ISeq:<compiled>@<compiled>:1 (1,0)-(3,3)> (catch: false)
0000 definemethod                           :foo, foo                 (   1)[Li]
0003 putobject                              :foo
0005 leave

== disasm: #<ISeq:foo@<compiled>:1 (1,0)-(3,3)> (catch: false)
local table (size: 1, argc: 0 [opts: 0, rest: -1, post: 0, block: 0, kw: -1@-1, kwrest: -1])
[ 1] block@0<Block>
0000 getblockparamproxy                     block@0, 0                (   2)[LiCa]
0003 branchunless                           11
0005 putself
0006 putobject                              :hey
0008 opt_send_without_block                 <calldata!mid:puts, argc:1, FCALL|ARGS_SIMPLE>
0010 leave                                                            (   3)[Re]
0011 putnil                                                           (   2)
0012 leave                                                            (   3)[Re]
```

Note the `getblockparamproxy` vs `getblockparam`.

Reply:
Is checking for `block` preferable over `block_given?`? I'm only reaching for `block` because of guidance from the linter.

Reply:
Not really, it's kinda the same. That linter rule I'm not fond of.